jiant/utils/retokenize.py

```diff
@@ -179,31 +181,91 @@ def project_tokens(self, idxs: Union[int, Sequence[int]]) -> Sequence[int]:
         """
         if isinstance(idxs, int):
             idxs = [idxs]
-        return self.T[idxs].nonzero()[1]  # column indices
-
-    def project_span(self, start, end) -> Tuple[int, int]:
+        return self.source_token_idx_to_target_token_idx[idxs].nonzero()[1]  # column indices
+
+    @classmethod
+    def _project_span(cls, mat, start, end, inclusive):
+        if inclusive:
+            end = end + 1
+        target_matches = mat[start:end].nonzero()[1].tolist()
+        if len(target_matches) == 0:
+            raise ValueError("Project into empty span in target sequence")
+        output_start, output_end = min(target_matches), max(target_matches)
+        if not inclusive:
+            output_end = output_end + 1
+        return (output_start, output_end)
+
+    def project_token_span(self, start, end, inclusive=False) -> Tuple[int, int]:
         """Project a span from source to target token sequence.
 
         Notes:
-            Span end is taken to be exclusive, so this actually projects end - 1 and maps back to an
-            exclusive target span.
+            If inclusive is False, span start is inclusive, while the end exclusive.
+            This affects input and output simultaneously.
 
         Examples:
             >>> source_tokens = ['abc', 'def', 'ghi', 'jkl']
             >>> target_tokens = ['abc', 'd', 'ef', 'ghi', 'jkl']
             >>> ta = TokenAligner(source_tokens, target_tokens)
             >>> start, end = 0, 2
-            >>> print(ta.project_span(start, end))
+            >>> print(ta.project_token_span(start, end))
             (0, 3)
 
-        Raises:
-            ValueError if span end index is not greater than start index.
+        Raise:
+            When target span is empty
 
         Returns:
-            Tuple[int, int] representing the target span (span end exclusive) for the source span.
+            Tuple[int, int] representing the target span corresponding to the source span.
+        """
+        return self._project_span(
+            mat=self.source_token_idx_to_target_token_idx, start=start, end=end, inclusive=inclusive
+        )
+
+    def project_token_to_char_span(self, start, end, inclusive=False) -> Tuple[int, int]:
+        """Project a span from source to target token sequence.
+
+        Notes:
+            If inclusive is False, span start is inclusive, while the end exclusive.
+            This affects input and output simultaneously.
```

Comment:
I recommend spelling out the input/output behavior, replacing "This affects input and output simultaneously." with something like...

"When param inclusive=False, the `end` index is interpreted as exclusive, and the span returned by the function will also be exclusive"... "When param inclusive=True, the...`

Reply:
(same note can/should be copied to `project_char_to_token_span` docstrings.)